File: solutions/java/code-carl/ch03_hashtable/T6_LC0454_FourSumCount.java
/**
 * Practice template for:
 * - LeetCode 454. 4Sum II
 *   https://leetcode.cn/problems/4sum-ii/
 */
import java.util.HashMap;

public class T6_LC0454_FourSumCount {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int res = 0;
        for (int n1 : nums1) {
            for (int n2 : nums2) {
                int sum = n1 + n2;
                map.put(sum, map.getOrDefault(sum, 0) + 1);
            }
        }
        for (int n3 : nums3) {
            for (int n4 : nums4) {
                int sum = n3 + n4;
                if (map.containsKey(-sum)) {
                    res += map.get(-sum);
                }
            }
        }

        return res;
    }
}

File: templates/python/code-carl/ch03_hashtable/t6_lc0454_four_sum_count.py
"""LeetCode 454. 4Sum II
https://leetcode.cn/problems/4sum-ii/
"""

from typing import List


class Solution:
    def fourSumCount(
        self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]
    ) -> int:
        pass


if __name__ == "__main__":
    solution = Solution()
    assert solution.fourSumCount([1, 2], [-2, -1], [-1, 2], [0, 2]) == 2
    assert solution.fourSumCount([0], [0], [0], [0]) == 1
    print("All tests passed.")
